metadata_store_scripts/V2025.07.22.0__add_generate_masking_parameters_procedure.sql

```diff
@@ -0,0 +1,391 @@
+ * - adf_type_mapping table: Maps database types to ADF types
+ * - Supports conditional masking via JSON structures in assigned_algorithm column
+ *
+ * FALLBACK BEHAVIOR:
+ * Always returns exactly one record. If no masking rules exist, returns default/empty
+ * values to prevent ADF lookup activity failures.
+ *
+ * NOTES:
+ * - Empty JSON arrays are converted to [""] for ADF UI compatibility
+ * - Uses hex-encoded column names to avoid ADF pipeline issues with special characters
+ * - Supports treat_as_string flag for type casting feature
+ */
+CREATE OR ALTER PROCEDURE generate_masking_parameters
+    @source_database NVARCHAR(128),
+    @source_schema NVARCHAR(128),
+    @source_table NVARCHAR(128),
+    @dataset NVARCHAR(128),
+    @column_width_estimate INT = 1000,
+    @filter_key NVARCHAR(128) = ''  -- Optional filter key for conditional masking
+AS
+BEGIN
+    SET NOCOUNT ON;
+
+    -- base_ruleset_filter - Filter to target table, add ADF type mappings
+    WITH base_ruleset_filter AS (
+        SELECT
+            r.*,
+            t.adf_type
+        FROM discovered_ruleset AS r
+        INNER JOIN adf_type_mapping AS t
+            ON
+                r.identified_column_type = t.dataset_type
+                AND r.dataset = t.dataset
+        WHERE
+            r.dataset = @dataset
+            AND r.specified_database = @source_database
+            AND r.specified_schema = @source_schema
+            AND r.identified_table = @source_table
+            AND r.assigned_algorithm IS NOT NULL
+            AND r.assigned_algorithm <> ''
+    ),
+
+    -- ruleset_computed - Add encoded column names and extract JSON metadata
+    ruleset_computed AS (
+        SELECT
+            r.*,
+            CONCAT(
+                'x',
+                CONVERT(
+                    VARCHAR(MAX),
+                    CONVERT(VARBINARY, r.identified_column),
+                    2
+                )
+            ) AS encoded_column_name,
+            JSON_VALUE(r.algorithm_metadata, '$.date_format') AS [date_format],
+            CONVERT(BIT, JSON_VALUE(r.algorithm_metadata, '$.treat_as_string')) AS treat_as_string,
+            CASE
+                WHEN ISJSON(r.assigned_algorithm) = 1
+                    THEN JSON_VALUE(r.assigned_algorithm, '$.key_column')
+            END AS [key_column_name]
+        FROM base_ruleset_filter AS r
+    ),
+
+    -- conditional_algorithm_extraction - Extract conditional algorithms for matching filter key
+    conditional_algorithm_extraction AS (
+        SELECT
+            r.*,
+            x.[key_column],
+            x.[conditions],
+            c.[condition_alias],
+            c.[algorithm] AS resolved_algorithm
+        FROM ruleset_computed AS r
+        CROSS APPLY
+            OPENJSON (r.assigned_algorithm)
+            WITH (
+                [key_column] NVARCHAR(255) '$.key_column',
+                [conditions] NVARCHAR(MAX) '$.conditions' AS JSON
+            ) AS x
```

Comment:
suggestion: pick a more meaningful abbreviation, `x` should perhaps instead be `aa` for `assigned_algorithm` or `alg`?